compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -558,72 +558,78 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        &inline::Inline,
+        // Code from other crates may have storage markers, so this needs to happen after inlining.
+        &remove_storage_markers::RemoveStorageMarkers,
+        // Inlining and instantiation may introduce ZST and useless drops.
+        &remove_zsts::RemoveZsts,
+        &remove_unneeded_drops::RemoveUnneededDrops,
+        // Type instantiation may create uninhabited enums.
+        // Also eliminates some unreachable branches based on variants of enums.
+        &unreachable_enum_branching::UnreachableEnumBranching,
+        &unreachable_prop::UnreachablePropagation,
+        &o1(simplify::SimplifyCfg::AfterUnreachableEnumBranching),
+        // Inlining may have introduced a lot of redundant code and a large move pattern.
+        // Now, we need to shrink the generated MIR.
+        &ref_prop::ReferencePropagation,
+        &sroa::ScalarReplacementOfAggregates,
+        &match_branches::MatchBranchSimplification,
+        // inst combine is after MatchBranchSimplification to clean up Ne(_1, false)
+        &multiple_return_terminators::MultipleReturnTerminators,
+        // After simplifycfg, it allows us to discover new opportunities for peephole optimizations.
+        &instsimplify::InstSimplify::AfterSimplifyCfg,
+        &simplify::SimplifyLocals::BeforeConstProp,
+        &dead_store_elimination::DeadStoreElimination::Initial,
+        &gvn::GVN,
+        &simplify::SimplifyLocals::AfterGVN,
+        &dataflow_const_prop::DataflowConstProp,
+        &single_use_consts::SingleUseConsts,
+        &o1(simplify_branches::SimplifyConstCondition::AfterConstProp),
+        &jump_threading::JumpThreading,
+        &early_otherwise_branch::EarlyOtherwiseBranch,
+        &simplify_comparison_integral::SimplifyComparisonIntegral,
+        &dest_prop::DestinationPropagation,
+        &o1(simplify_branches::SimplifyConstCondition::Final),
+        &o1(remove_noop_landing_pads::RemoveNoopLandingPads),
+        &o1(simplify::SimplifyCfg::Final),
+        &copy_prop::CopyProp,
+        &dead_store_elimination::DeadStoreElimination::Final,
+        &nrvo::RenameReturnPlace,
+        &simplify::SimplifyLocals::Final,
+        &multiple_return_terminators::MultipleReturnTerminators,
+        &deduplicate_blocks::DeduplicateBlocks,
+        &large_enums::EnumSizeOpt { discrepancy: 128 },
+        // Some cleanup necessary at least for LLVM and potentially other codegen backends.
+        &add_call_guards::CriticalCallEdges,
+        // Cleanup for human readability, off by default.
+        &prettify::ReorderBasicBlocks,
+        &prettify::ReorderLocals,
+        // Dump the end result for testing and debugging purposes.
+        &dump_mir::Marker("PreCodegen"),
+    ];
+
+    let mut passes = passes.to_vec();
+    let def_id = body.source.def_id();
+    if tcx.def_kind(def_id).has_codegen_attrs()
+        && let Some(ref attrs) = tcx.codegen_fn_attrs(def_id).optimize
+        && attrs.do_not_optimize()
```

Comment:
Add a comment here explaining what you are doing, like "// if the `#[optimize(none)]` attribute is set, remove any pass that isn't opt level 0."